src/overlays/actors/ovl_En_Niw/z_en_niw.c

```diff
@@ -21,61 +48,1155 @@ const ActorInit En_Niw_InitVars = {
+                     EnNiw_OverrideLimbDraw, NULL, &this->actor);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB70F8.s")
+    if (this->actionFunc == func_80AB6450) {
+        func_80033C30(&this->actor.posRot.pos, &scale, 255, globalCtx);
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB714C.s")
+    EnNiw_FeatherDraw(this, globalCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7204.s")
+void EnNiw_FeatherSpawn(EnNiw* this, Vec3f* pos, Vec3f* vel, Vec3f* accel, f32 scale) {
+    s16 i;
+    EnNiwFeather* feather = this->feathers;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7290.s")
+    for (i = 0; i < ARRAY_COUNT(this->feathers); i++, feather++) {
+        if (feather->type == 0) {
+            feather->type = 1;
+            feather->pos = *pos;
+            feather->vel = *vel;
+            feather->accel = *accel;
+            feather->timer = 0;
+            feather->scale = scale / 1000.0f;
+            feather->lifespan = (s16)Math_Rand_ZeroFloat(20.0f) + 40;
+            feather->unk_2A = Math_Rand_ZeroFloat(1000.0f);
+            break;
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7328.s")
+void EnNiw_FeatherUpdate(EnNiw* this, GlobalContext* globalCtx) {
+    s16 i;
+    EnNiwFeather* feather = this->feathers;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7420.s")
+    for (i = 0; i < ARRAY_COUNT(this->feathers); i++, feather++) {
+        if (feather->type != 0) {
+            feather->timer++;
+            feather->pos.x += feather->vel.x;
+            feather->pos.y += feather->vel.y;
+            feather->pos.z += feather->vel.z;
+            feather->vel.x += feather->accel.x;
+            feather->vel.y += feather->accel.y;
+            feather->vel.z += feather->accel.z;
+            if (feather->type == 1) {
+                feather->unk_2A++;
+                Math_SmoothScaleMaxF(&feather->vel.x, 0.0f, 1.0f, 0.05f);
+                Math_SmoothScaleMaxF(&feather->vel.z, 0.0f, 1.0f, 0.05f);
+                if (feather->vel.y < -0.5f) {
+                    feather->vel.y = -0.5f;
+                }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB747C.s")
+                feather->unk_30 = Math_Sins(feather->unk_2A * 0xBB8) * M_PI * 0.2f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/EnNiw_Update.s")
+                if (feather->lifespan < feather->timer) {
+                    feather->type = 0;
+                }
+            }
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Niw/func_80AB7F60.s")
+void EnNiw_FeatherDraw(EnNiw* this, GlobalContext* globalCtx) {
+    u8 flag = 0;
+    s16 i;
+    s32 pad;
+    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
+    EnNiwFeather* feather = this->feathers;
```

Comment:
is the supposed to be a pointer to the first feather in the array? if so then it should probably be `feather = &this->feathers[0];`